### ISSUE TITLE
Checkstyle configuration from URI

### ISSUE THREAD

**Issue by filiphr** [gradle/gradle#2663]
Currently, there is no way to configure checkstyle to use a configuration from a given URL (at least not in a clean way and in a way that I am aware of). 

The idea is to have a checkstyle configurations for an organization that can be hosted on one location and other projects within that organization can use that configuration file in multiple projects without the need to copy it.

This can be something similar as the `headerURI` from the [gradle-license-plugin](https://github.com/hierynomus/license-gradle-plugin)

**Comment by oehme** [comment]
The [setConfig](https://docs.gradle.org/current/javadoc/org/gradle/api/plugins/quality/Checkstyle.html#setConfig(org.gradle.api.resources.TextResource)) method accepts a `TextResource`, which can also refer to a File through a URI.

**Comment by filiphr** [comment]
@oehme how can I actually do that? The file I need is prefixed with `http`. (it is actually in another github repository)

If your suggestion is for me to download the file and then create a file from it and pass it to the `config` that is what I wanted to avoid doing.

**Comment by oehme** [comment]
You just pass the URI to `project.resources.text.fromFile()`

**Comment by filiphr** [comment]
That is what I've tried and it is not working:

My config looks like:

```gradle
checkstyle {
    config project.resources.text.fromFile(new URI("https://raw.githubusercontent.com/mapstruct/mapstruct/master/build-config/src/main/resources/build-config/checkstyle.xml" ) )
}
```

I've also tried with just the string and calling `toUri()` on it.

I am getting this exception:

```
FAILURE: Build failed with an exception.
* What went wrong:
Cannot convert URL 'https://raw.githubusercontent.com/mapstruct/mapstruct/master/build-config/src/main/resources/build-config/checkstyle.xml' to a file.
* Try:
Run with --info or --debug option to get more log output.
* Exception is:
org.gradle.api.InvalidUserDataException: Cannot convert URL 'https://raw.githubusercontent.com/mapstruct/mapstruct/master/build-config/src/main/resources/build-config/checkstyle.xml' to a file.
	at org.gradle.api.internal.file.AbstractFileResolver.convertObjectToFile(AbstractFileResolver.java:124)
	at org.gradle.api.internal.file.BaseDirFileResolver.doResolve(BaseDirFileResolver.java:76)
	at org.gradle.api.internal.file.AbstractFileResolver.resolve(AbstractFileResolver.java:81)
	at org.gradle.api.internal.file.AbstractFileResolver.resolve(AbstractFileResolver.java:63)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext$FileCollectionConverter.convertInto(DefaultFileCollectionResolveContext.java:174)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext.doResolve(DefaultFileCollectionResolveContext.java:135)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext.resolveAsFileCollections(DefaultFileCollectionResolveContext.java:93)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext$FileCollectionConverter.convertInto(DefaultFileCollectionResolveContext.java:158)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext.doResolve(DefaultFileCollectionResolveContext.java:110)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext.resolveAsFileCollections(DefaultFileCollectionResolveContext.java:93)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext$FileCollectionConverter.convertInto(DefaultFileCollectionResolveContext.java:158)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext.doResolve(DefaultFileCollectionResolveContext.java:110)
	at org.gradle.api.internal.file.collections.DefaultFileCollectionResolveContext.resolveAsFileCollections(DefaultFileCollectionResolveContext.java:93)
	at org.gradle.api.internal.file.CompositeFileCollection.getSourceCollections(CompositeFileCollection.java:172)
	at org.gradle.api.internal.file.CompositeFileCollection.visitRootElements(CompositeFileCollection.java:184)
	at org.gradle.api.internal.changedetection.state.AbstractFileCollectionSnapshotter.snapshot(AbstractFileCollectionSnapshotter.java:51)
	at org.gradle.api.internal.changedetection.state.DefaultGenericFileCollectionSnapshotter.snapshot(DefaultGenericFileCollectionSnapshotter.java:36)
	at org.gradle.api.internal.changedetection.rules.AbstractNamedFileSnapshotTaskStateChanges.buildSnapshots(AbstractNamedFileSnapshotTaskStateChanges.java:91)
	at org.gradle.api.internal.changedetection.rules.AbstractNamedFileSnapshotTaskStateChanges.<init>(AbstractNamedFileSnapshotTaskStateChanges.java:58)
	at org.gradle.api.internal.changedetection.rules.InputFilesTaskStateChanges.<init>(InputFilesTaskStateChanges.java:29)
	at org.gradle.api.internal.changedetection.rules.TaskUpToDateState.<init>(TaskUpToDateState.java:60)
	at org.gradle.api.internal.changedetection.changes.DefaultTaskArtifactStateRepository$TaskArtifactStateImpl.getStates(DefaultTaskArtifactStateRepository.java:178)
	at org.gradle.api.internal.changedetection.changes.DefaultTaskArtifactStateRepository$TaskArtifactStateImpl.isUpToDate(DefaultTaskArtifactStateRepository.java:88)
	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:53)
	at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:242)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:95)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:235)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:224)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:121)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:77)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:102)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:96)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:612)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:567)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:96)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
```

I thought that it is not allowed to do what I wanted to do, but you just said that passing URI to the `fromFile` should work, so I am pasting the exception as well.

**Comment by oehme** [comment]
You're right, only file:/ urls are supported, which doesn't help in your case. What you could do is download the file in another task as a workaround.

**Comment by filiphr** [comment]
Thanks for reopening this. 

I wanted to avoid doing the workaround and that is why I posted the issue. The workaround you said is what I meant in a clean way 😀. 

In any case are you open in adding a new parameter to the configuration or you want to solve it in a different way?

**Comment by big-guy** [comment]
There used to be a blog post that described exactly how to do this... but we seem to have lost it somewhere along the way.

The idea was to put your checkstyle (or all of your code quality config, really) into a zip that could be downloaded:

```
configurations {
   codeQuality
}

dependencies {
   codeQuality 'com.example:codeQuality:1.0@zip'
}
```

And then in the Checkstyle and other code quality tasks, you could configure them with:
```
checkstyle {
    config project.resources.text.fromArchiveEntry(configurations.codeQuality, "build-config/checkstyle.xml")
}
```

But I like the idea of being able to do `resources.text.fromUrl(...)` a lot (especially if we cache/only download it when needed).

**Comment by big-guy** [comment]
Aha, internet archive to the rescue:

http://web.archive.org/web/20150322034610/http://www.gradleware.com/tutorials/feature-spotlight-enforcing-code-quality-standards/

**Comment by filiphr** [comment]
Thanks for the link, that's a good one. In my case it won't work as we have a different way of doing it, but the idea is that. 

And yes having a `resources.text.fromUrl(...)` would be great (with the addition of caching, here one can use the HTTP cache, etag headers etc).

**Comment by oehme** [comment]
We already have url caching for `apply from:`, so adding this feature could be a cool contribution. Would you be interested in submitting a PR?

**Comment by filiphr** [comment]
I've never looked at the gradle codebase, but it would be interesting to try it out 😄 . And I've just recently tried working with gradle on an IntelliJ plugin. In any case I can give it a try, why not 😄 .

Just to make sure I understand correctly, I'll have to work with the [code-quality](https://github.com/gradle/gradle/tree/master/subprojects/code-quality) subproject, more specifically the `CheckstyleException`. Should I add a new config parameter for this or we would need to somehow have a `TextResource` that can be created from an URI.

Where is the `apply from:` located so I can have a look at it?

**Comment by oehme** [comment]
You'd extend the [TextResourceFactory](https://github.com/gradle/gradle/blob/master/subprojects/core-api/src/main/java/org/gradle/api/resources/TextResourceFactory.java) to have `fromUri`. The `apply from:` code is handled by [UriTextResource](https://github.com/gradle/gradle/blob/master/subprojects/resources/src/main/java/org/gradle/internal/resource/UriTextResource.java). Note that we have two `TextResource` interfaces, one on the public API and one internally. You can probably reuse most of the internal stuff to implement the new public method.

**Comment by filiphr** [comment]
Thanks for the pointers @oehme. I'll clone the repo and have a look at the source. I'd extend the factory and go from there. 

I'll try to see if I can write tests first so I can get the hang of it easier 😄 .

Btw, I think you need to change the label of the issue, it is not really a question anymore, or?

**Comment by filiphr** [comment]
@oehme Should we do something else before i start working on this?

**Comment by oehme** [comment]
I think we don't need a big design for this, as the scope is pretty clear and the semantics would be the same as the existing `apply from:`. Starting with tests is a good plan, they are required for accepting the PR in any case :)

**Comment by filiphr** [comment]
Great. Can you perhaps point me to the implementation of `apply from:`? 

By the way the internal `UriTextResource` is not cacheable. 

And do you have a gitter or something like that so I don't spam the issue here in case of some questions :smile:

**Comment by oehme** [comment]
`apply from:` is handled by `ObjectConfigurationAction`. And you are right, my first pointer was wrong, because that's the old, uncached uri resource it used to use (and which is still the fallback for some corner cases). See `TextResourceLoader` instead.

**Comment by filiphr** [comment]
Thanks for the pointers @oehme it helped me a lot. I had a look and I think I know how to solve, I just need some more help 😄 .

I saw that the "heavy" lifting for loading an internal `TextResource` is done by the `DefaultUriTextResourceLoader` and then by the `ExternalResourceAccessor` (which I think does the caching), or maybe even `CacheAwareExternalResourceAccessor` is responsible for the caching. I would like to use the `DefaultUriTextResourceLoader` to load the URI and convert it into an api `TextResource`. Is this a good approach? How can I create a `DefaultUriTextResourceLoader`.

If that is a good approach, then I would use the internal `TextResource` to get the file from it (or the reader and create a temporary file from the reader). However, I think that I might have found a bug there. The `DownloadedUriTextResource` has the downloaded resource passed to it and the only place where that is used it during `openReader`. However, the `getFile` method there returns the `sourceFile` and not the `downloadedResource`  is this correct? 

I also have a question about the tests, if I go with using the `DefaultUriTextResourceLoader` then I can easily mock the call and return a locally created file. However, if that is not the case then I will probable need to

**Comment by oehme** [comment]
There is no need to create any of these services, you simply inject the `TextResourceLoader` into the constructor of the class that needs it (in this case the `DefaultTextResourceFactory`). You'll have a few layers of indirection above you (`DefaultResourceHandler` and `DefaultFileOperations`), which takes you to the place where the [service registration happens](https://github.com/gradle/gradle/blob/f1efee61bcee87411f7b78761cbb492250e03b70/subprojects/core/src/main/java/org/gradle/internal/service/scopes/ProjectScopeServices.java#L143).

Regarding testing, it'll be even easier since you just inject an interface. Please also make sure to add a functional test which actually runs an end-to-end build (there are many examples in the code base). We have http server test fixtures that can simulate a remote resource.

**Comment by filiphr** [comment]
@big-guy I've created PR #2760. With the PR there is support for `TextResources` backed by an URI.

However, please have a look at my comment regarding my reasoning why I think that the solution will not work for my initial request.

**Comment by oehme** [comment]
This has been merged and will be in 4.8

**Comment by michapringle** [comment]
TL;DR; You can use this

``` 
task checkstyle(type: Checkstyle) {
          // from file
          configFile = file("${rootProject.projectDir}/checkstyle.xml")
          // from url
          resources.text.fromUri("${someRepoUrl}/checkstyle-0.1.0.xml")
}
```

**Comment by the4yeast** [comment]
is there a way to pass headers to `fromUri`? that would enable fetching config from private repos.
